[Enhancement] add seedsigner_gettext with improved variable handling

AdityaPandeyCN opened the pull request:
## Description

_This PR addresses issue #693 by implementing enhanced gettext functions with better error handling for translation variables.
When translations are missing named variables (like when German has {} instead of {mnemonic_length}), the current error message is unhelpful:_

`IndexError: Replacement index 0 out of range for positional args tuple`

My implementation provides clear error messages that specifically identify the issue:

```
Translation error in locale 'de': Translation missing variables {mnemonic_length}, {num_bits}.
Original: 'The {mnemonic_length}th word is built from {num_bits} more entropy bits plus auto-calculated checksum.'
Translation: '{}es Wort wird erstellt aus {} mehr Entropiebits plus errechneter Prüfsumme'
```
Changes:
1. Added` seedsigner_gettext` and `seedsigner_ngettext `functions with variable handling
2. Added `TranslationVariableMissingError` class with detailed error information
3. Updated imports across the codebase to use the enhanced functions
 
This helps translators quickly identify and fix variable formatting issues.

This pull request is categorized as a:

- [x] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Documentation
- [ ] Other

## Checklist

- [x] I’ve run `pytest` and made sure all unit tests pass before sumbitting the PR

If you modified or added functionality/workflow, did you add new unit tests?

- [ ] No, I’m a fool
- [ ] Yes
- [x] N/A

I have tested this PR on the following platforms/os:

- [ ] Raspberry Pi OS [Manual Build](https://github.com/SeedSigner/seedsigner/blob/dev/docs/manual_installation.md)
- [ ] [SeedSigner OS](https://github.com/SeedSigner/seedsigner-os) on a Pi0/Pi0W board
- [x] Other

kdmukai committed:
rough prototype

kdmukai committed:
Tweaks to SeedMnemonicEntryScreen layout

kdmukai committed:
Expand xpub in SeedExportXpubDetailsScreen

kdmukai committed:
new `ToolsAddressExplorerAddressListScreen`

kdmukai committed:
Screensaver relative positioning

Still seems a bit off (overshoots the horizontal center to the right), but close enough.

kdmukai committed:
basic ili9341 working, but no portrait only

kdmukai committed:
Support for new IPS display

kdmukai committed:
Starting to generalize support for the different displays

kdmukai committed:
Update settings_definition.py

kdmukai committed:
on-the-fly display driver changes via SettingsQR

kdmukai committed:
Update settings_definition.py

kdmukai committed:
Update st7789_mpy.py

kdmukai committed:
compatibility for st7789 240x320; re-init display on settings change

alvroble committed:
Enhancement for issue #546

Provides 3 microSD options:
- Toast disabled
- 5-sec toast (default)
- SD card removal required* (fullscreen dialog,
cannot proceed unless microSD is removed)

- "SD card removal required" option: warning view
shown instead of fullscreen toast so user can
disable this setting without removing SD card.

alvroble committed:
Updates test suite

newtonick committed:
added support for BBQr decoding support

newtonick committed:
add unit tests for bbqr and fix deprecation warning for invalid escape sequence '$'

fedebuyito committed:
Update encode_qr.py

Some helpers comments.

fedebuyito committed:
Update seed_views.py

Cleanup the else condition so that the return to SeedTranscribeSeedQRConfirmInvalidQRView occurs on all remaining possibilities (any QR that is not SeedQR).

fedebuyito committed:
Update test_flows_seed.py

Includes additional failure condition (when QR is recognized but is not a SeedQR) in flow test, directed toward SeedTranscribeSeedQRConfirmInvalidQRView

kdmukai committed:
Hard-code lots of locales now to enable autodiscovery later

kdmukai committed:
Variant to drill into l10n dirs and confirm *.mo

kdmukai committed:
Remove unused fonts

kdmukai committed:
Update `coverage` to 7.6.12

Commit:
only decrement cursor position when character is removed

kdmukai committed:
incorporating changes to support Asian languages, other improvements

kdmukai committed:
Adding missing fonts

kdmukai committed:
Slight enhancement to `coverage` calcs

PROWLERx15 committed:
Modify CI to run on all commits

krishnavbajoria02 committed:
Docs: Clarify that SeedQR uses English BIP39 wordlist

AdityaPandeyCN committed:
add seedsigner_gettext with improved variable handling and error messages

Signed-off-by: AdityaPandeyCN <adityapand3y666@gmail.com>

Chaitanya-Keyal committed:
docs: add note about `--ignore-missing` flag issue on older macOS versions

PROWLERx15 committed:
removed branches block to make it cleaner

krishnavbajoria02 committed:
Added 'calculating' spinner

Advaitgaur004 committed:
fix: Update capitalization to Title Case for titles and Sentence case for buttons (#648)

Advaitgaur004 committed:
fix: Update UI to Sentence case for buttons.

- Updated buttons to Sentence case, capitalizing only the first word.
- Preserved proper nouns (SeedQR, BIP-39 Passphrase, Native Segwit, Nested Segwit, Taproot).
- Ensured Xpub and Custom derivation are not always capitalized.

PROWLERx15 committed:
update docs

PROWLERx15 committed:
fix typo

megha-66 committed:
Tab spacing to Spaces

FazleRabbbiferdaus172 committed:
[IMP] Disable screensaver during SeedQR transcription

Disables the screensaver entirely while the SeedQR transcription routine is active. Any view can be configured to allow or disallow screensaver activation.

kdmukai committed:
Bugfixes

kdmukai committed:
Restore original ST7789 driver for 240x240

kdmukai committed:
Use checkmark to indicate selected language, ensure initial scroll displays it

kdmukai committed:
Fix centering on OpeningSplashScreen for 320x240

kdmukai committed:
Aspect ratio-savvy camera frame resizing for non-square displays

FazleRabbbiferdaus172 committed:
[FIX] active_view returns MainMenuView when back stack is empty

FazleRabbbiferdaus172 committed:
[IMP] can_run_screensaver implemented to handle screensaver runable logics

FazleRabbbiferdaus172 committed:
[IMP] code cleanup

kdmukai committed:
Bugfix: SeedQR transcription UI navigation; cleanup

kdmukai committed:
simplified, optimized SeedQR transcription mask overlay rendering

kdmukai committed:
Finish integrating ToolsAddressExplorerAddressListScreen

kdmukai committed:
standardize on our landscape orientation (widest dim first)

kdmukai committed:
Fix up one last value

kdmukai committed:
bugfix after swapping st7789 dims

kdmukai committed:
Restrict pre-scrolling ButtonList logic to just long lists

FazleRabbbiferdaus172 committed:
[REF] active_view property removed and can_start_screensaver renamed to is_screensaver_start_allowed

kdmukai committed:
Add `invert` to `ST7789` driver

kdmukai committed:
Add "beta" designation to ili9341 driver

kdmukai committed:
Reorder Settings screenshots, remove 320x240 hard coding

kdmukai commented:
I have not yet carefully reviewed the changes, but this PR **definitely** needs accompanying test cases for all of the error AND expected success conditions of the enhanced `gettext`.

See the test case pattern that goes something like:
```python
with pytest.raises(SomeException):
    # test code that intentionally raises an Exception...
```

kdmukai committed:
bugfix: don't set class attr to dynamic method

fedebuyito committed:
Update scan_screens.py

Added a return value in the _run() method to evaluate whether the back keys are pressed when a QR code is scanned.

fedebuyito committed:
Update seed_views.py

Added a return to the last view in case the back button has been pressed in the SeedTranscribeSeedQRConfirmScanView class.

fedebuyito committed:
Update seed_views.py

Typo corrected

FazleRabbbiferdaus172 committed:
[Temporary] rough test included that test the screensaver start status when the current view is SeedTranscribeSeedQRZoomedInView

FazleRabbbiferdaus172 committed:
[Temporary] TODO added

kdmukai committed:
Handle PiCameraError more gracefully

kdmukai committed:
Fix error screen button text

kdmukai committed:
Explicitly add CameraConnectionError as our UnhandledExceptionView screenshot

kdmukai committed:
Remove unnecessary import

kdmukai committed:
Add custom error View for `CameraConnectionError`

kdmukai committed:
Add flow test for `CameraConnectionErrorView`

kdmukai committed:
remove unused import

kdmukai committed:
Improved flow test exception simulation

alvroble committed:
Remove 200ms sleep in toast rendering

kdmukai committed:
Proper detection of chars rendering below baseline on final line of text

AdityaPandeyCN commented:
Hello, apologies for late reply was stuck with exams. I will soon add test for this, Thankyou

tadeubas committed:
use pypi version instead of commit for urtypes lib

kdmukai committed:
Promote zh_Hans_CN to beta

kdmukai committed:
Marks new display config settings text for translation

kdmukai committed:
Add translator notes

kdmukai committed:
bugfix

kdmukai committed:
[l10n] ToolsCalcFinalWordScreen: adjusts spacing for locales w/larger fonts

Chaitanya-Keyal committed:
fix: restart on manual build

Chaitanya-Keyal committed:
docs: ensure `Restart=no` in manual build

Chaitanya-Keyal committed:
fix: use `os.execv` for restart

kdmukai committed:
Remove `TextArea.allow_text_overflow`

kdmukai committed:
Add screenshot for brightness tips during QR display

kdmukai committed:
Comment update, cleanup

kdmukai committed:
Move definition out of function

Advaitgaur004 committed:
fix: Update capitalization for "Electrum" seed option in multiple views

Chaitanya-Keyal committed:
fix(`RestartView`): stop restarting for screenshot renderer

kdmukai committed:
Exhaustively test `PSBTParser` with all supported script types

kdmukai committed:
clarify comment

kdmukai committed:
minor import fix

kdmukai committed:
Adds Arabic font, enables Arabic and Persian

PROWLERx15 committed:
cleaned bip references to BIP-XX format

alvroble committed:
Add libraqm in dev build instutcions

alvroble committed:
Remove "libraqm not supported" tweaks

PROWLERx15 committed:
changes

bitcoinprecept committed:
Update controller.py

Version bump

bitcoinprecept committed:
Update pyproject.toml

newtonick committed:
update seedsigner-translations submodule to latest version

newtonick committed:
update submodule seedsigner-translations to latest after adding back compiled mo files for 0.8.6 release

PROWLERx15 committed:
removed

PROWLERx15 committed:
changed

kdmukai committed:
Move Italian to the "Fully supported languages" list

kdmukai committed:
Exclude hardware drivers from `coverage`

kdmukai committed:
Use more explicit path

kdmukai committed:
Promote Thai to beta status; add Thai font

SeedSigner committed:
Update README.md

Update instances of 0.8.5 to 0.8.6

newtonick committed:
remove telegram notification github action since it's not working

alvroble committed:
Add logger message if libraqm not installed

notTanveer committed:
fix(BIP-39 passphrase): fallback to ExitDialog, if empty passphrase

notTanveer committed:
always route to exit dialog, if empty and < button is pressed

alvroble committed:
Update src/seedsigner/controller.py

Co-authored-by: notTanveer <101289209+notTanveer@users.noreply.github.com>

notTanveer committed:
imporove exit dialog UX

Chaitanya-Keyal committed:
fix(restart): dont start thread for screenshot renderer

notTanveer committed:
fix if-else conditionals

notTanveer committed:
fix: typo

alvroble committed:
Refactor to enable auto-detection of screenshot generation renderer

FazleRabbbiferdaus172 committed:
Seedsigner 2 lines convention for method

FazleRabbbiferdaus172 committed:
doc string update

FazleRabbbiferdaus172 committed:
backstack comment update

FazleRabbbiferdaus172 committed:
property renamed as per suggestion

FazleRabbbiferdaus172 committed:
todo removed

PROWLERx15 committed:
follow BIP-XX format

PROWLERx15 committed:
fix dice roll count

PROWLERx15 committed:
changes

Chaitanya-Keyal committed:
refactor: replace `.display()` with `View.run_screen`

Advaitgaur004 committed:
Update src/seedsigner/gui/screens/psbt_screens.py

Co-authored-by: notTanveer <101289209+notTanveer@users.noreply.github.com>

FazleRabbbiferdaus172 committed:
replace PSBT text in codebase with Proposed transaction or Transaction

PSBT text replaced with Proposed Transaction

FazleRabbbiferdaus172 committed:
[FIX] generate png for restart view

run_screen need to be executed by screenshot generator test to generate png for restart view. But we don't want to restart so the not starting the thread should be the way.

FazleRabbbiferdaus172 committed:
log moved to DoResetThread

Placing the log message inside the thread means it gets written at the exact moment the restart process actually begins. As in case of restart view this never gets executed. Better encapsulates the component responsible for rest.

FazleRabbbiferdaus172 committed:
explicit file regeneration check to screenshot generator

Previously, there were no checks to determine if a file was correctly generated by the generator. It was even hard to detect because of previously generated files.

kdmukai committed:
Add `ScreenshotRenderer` check to catch missing screenshots

kdmukai committed:
Remove unnecessary: toast calls, `ScreenshotComplete` reraise

kdmukai committed:
remove unnecessary blank line

FazleRabbbiferdaus172 committed:
Revert "explicit file regeneration check to screenshot generator"

This reverts commit e32604c1a60668079b207c0f00c9879ae5dfb651.

FazleRabbbiferdaus172 committed:
[REF] __post_init__ removed and RestartView initiates and start inside run method

FazleRabbbiferdaus172 committed:
[REF] logger in is moved above the comment

PROWLERx15 committed:
DRY refactor

Chaitanya-Keyal committed:
refactor: remove bip85 app num

FazleRabbbiferdaus172 committed:
Sentence case update for buttons

alvroble committed:
Use is_screenshot_generator for RestartView

Chaitanya-Keyal committed:
fix: add `ButtonOptionWithoutTranslation`

alvroble committed:
Add PSBT with empty fingerprint support

* Added `fill_zero_fingerprints` and `_fill_zero_fingerprint_scope` to fill
zero fingerprints each time before parsing a PSBT.
* Modified `has_matching_input_fingerprint` to match also on zero fingerprint
PSBTs and remove the (?) sign for matching seeds.

alvroble committed:
Add test case test_zero_fingerprint_handling

newtonick commented:
@AdityaPandeyCN there is interest in moving this PR forward, can you add a test like you described in May?

AdityaPandeyCN commented:
Hello @newtonick I will need more time on this as I have to do the project setup again due to an unfortunate system formatting. I  would be grateful if someone wants to pick up from here or I may need some time.

kdmukai committed:
Improve clarity on BIP-39 wordlist and indexing

Clarify language and indexing details in README.

kdmukai committed:
minor edits

kdmukai committed:
Restore minor button text change

kdmukai committed:
Import numpy, pivideostream in background

kdmukai committed:
Minor bugfix; mocks for test suite

newtonick committed:
remove foreign language fonts copied to seedsigner-translations

newtonick committed:
modify Fonts Class and get_font function to load fonts from 2 possible locations

newtonick committed:
update seedsigner-translations submodule to commit with foreign language font files

alvroble committed:
Option MICROSD_TOAST_TIMER_FOREVER refactor

- Prevent escaping the RemoveMicroSDWarningView by launching SettingsEntryUpdateSelectionView
  in a blocking mode. If the tracked setting remains unchanged, Back/normal navigation
  returns to the blocking view (RemoveMicroSDWarningView).
- Only 2 ways of escaping this mode: 1) physically removing MicroSD and
2) modifying the setting
- Add test tests/test_flows.py::TestFlowTest::test_remove_microsd_blocking covering:

kdmukai committed:
Refactor `DisplayDriver` to be a true parent class; add Factory

kdmukai committed:
Rename to `BaseDisplayDriver`; ili9341 changes

kdmukai committed:
Updating `messages.pot` to `dev` at `3455aa4`

kdmukai committed:
Updating `messages.pot` to `dev` at `3455aa4`

kdmukai committed:
Minor fixes to title case and consistency

kdmukai committed:
Exception str does not need to be translated

kdmukai committed:
Remove unreachable code

kdmukai committed:
Light reorg and simplification of `PSBTChangeDetailsScreen`

kdmukai committed:
Capitalization fix for button option

kdmukai committed:
add missing screenshots for single sig change verification

kdmukai committed:
Button text change, added TODO comment

alvroble committed:
Improve naming, comments and code consistency
* Added explanatory comments and reference to the original issue
* Better variable naming and type hinting to increase clarity
* Modified nested functions handling, naming and (intended) visibility

alvroble committed:
Improvement of tests, single-sig isolation
* Better naming and code efficiency
* Removed multisig from test inputs until figured out

alvroble committed:
Fix multisig tests

alvroble committed:
Improve commenting and solve minor nits

kdmukai committed:
Remove dead code

kdmukai committed:
Remove dead code, unused constants

kdmukai committed:
Remove dead code

kdmukai committed:
Remove dead code

kdmukai committed:
Remove dead code and debugging code

kdmukai committed:
Remove dead code

kdmukai committed:
Remove dead code

kdmukai committed:
Remove now-unused `has_passphrase` attr

kdmukai committed:
Replace the "coordinator" setting with "xpub_qr_type"

kdmukai committed:
Initial Hindi support

securesigner committed:
style: fix typos in source code comments and strings

kdmukai committed:
Remove sleeps that cause delays exiting thread

kdmukai committed:
Eliminate unnecessary scrolling for short text

kdmukai committed:
Relocate logic for min scroll threshold

kdmukai committed:
Cleanup

kdmukai committed:
Refactor scrolling loop for better clarity; more comments

kdmukai committed:
Light loop refactor; improved comments

kdmukai committed:
Remove last "coord" Settings references

kdmukai committed:
UX changes to require a selection for multiselect settings

kdmukai committed:
SettingsQR values cannot be blank

kdmukai committed:
Ensure persistent settings load valid multiselects

kdmukai committed:
Minor cleanup

Bicaru20 committed:
Eliminate Export xpub settings

Bicaru20 committed:
Changing tests

Change tests

Correcting tests

kdmukai committed:
Bugfix: `Mock` persisted beyond intended test case

Bicaru20 committed:
Add a recognized setting for test_settingsqr_ignores_unrecognized_setting

Bicaru20 committed:
Correcting assert mistake

kdmukai committed:
Fix inadvertantly persistent test artifacts

kdmukai committed:
Better bulletproofing

kdmukai committed:
Minor test improvement

kdmukai committed:
Remove unnecessary patch

kdmukai committed:
Remove unnecessary error message

kdmukai committed:
Add `mock_context_manager`; factor out all `run_before` methods

kdmukai committed:
Minor comments updates

kdmukai committed:
Improvements for SettingsQR Generator

kdmukai committed:
Improve tests, maintainability

kdmukai committed:
Remove duplicated attr entry in test

Chaitanya-Keyal committed:
fix(l10n): mark missing strings for translation

Chaitanya-Keyal committed:
fix(l10n): use named variables in `NetworkMismatchView` message

Co-authored-by: kdmukai <934746+kdmukai@users.noreply.github.com>

Chaitanya-Keyal committed:
fix(l10n): update MicroSD toast timer terminology and add translator notes

Chaitanya-Keyal committed:
chore(l10n): regenerate `messages.pot`

kdmukai committed:
bufixes: dataclass; param

kdmukai committed:
Restore whitespace alignment formatting

alvroble committed:
Bump Pillow version to 10.3.0

kdmukai committed:
Update PR template

kdmukai committed:
Explicit description template; minor presentation reformatting; docs link fix

Chaitanya-Keyal committed:
fix(IOTestScreen): stop blanking out "Clear" button for the screenshot renderer

Co-authored-by: kdmukai <934746+kdmukai@users.noreply.github.com>

Chaitanya-Keyal committed:
refactor(l10n): make multisig wallet policy display translatable

Chaitanya-Keyal committed:
refactor(l10n): update translator notes for multisig policy display

newtonick committed:
minor comment updates

newtonick committed:
update seedsigner-translations submodule to latest dev

Chaitanya-Keyal committed:
chore: bump version to 0.8.7

kdmukai committed:
Finalize v0.8.7 languages

kdmukai committed:
Add missing "beta" designation for Czech

S1DDHEY committed:
added tests for seedsigner_gettext error handling

S1DDHEY commented:
Hello @AdityaPandeyCN I've added tests for this which I've also tested locally. I currently don't have write access to this branch so I cannot push my code. Can you add me as a collaborator in your fork of this repo so that I can then push my changes directly to your branch instead of creating a separate PR

AdityaPandeyCN commented:
That will be bit confusing, just build a new PR on top of this.

S1DDHEY committed:
added *args support, migrate seed_screens call site, added positional arg tests

S1DDHEY commented:
Hi @AdityaPandeyCN , as per mentioned by Keith in https://github.com/SeedSigner/seedsigner/pull/910#issuecomment-4230522326 I should first try pushing to the original branch but as I don't have the write access, I cannot do that. So can you provide me the push access (i.e. adding me as a collaborator on your fork) or should I just keep the new PR #910 open

AdityaPandeyCN commented:
I have added you as collaborator @S1DDHEY

kdmukai commented:
Just for clarification, you can target a PR to anyone's fork and any branch within that fork. The fork owner then has the merge decision about the change.

So rather than giving permission to someone else within your own fork, which I would view as generally a bad idea, you could instead do:

* @S1DDHEY creates a PR to propose adding their `S1DDHEY:enhance-l10n-error-handling` changes to `AdityaPandeyCN:enhance-l10n-error-handling` branch.

* @AdityaPandeyCN reviews the PR and merges it into their `enhance-l10n-error-handling` branch.

* That would result in this PR #702 being automatically updated with S1DDHEY's changes.

---

<img width="1504" height="603" alt="Screenshot 2026-04-12 at 8 24 38 AM" src="https://github.com/user-attachments/assets/42770e51-dd4a-45fc-a0d1-e086ed62bdb5" />

I couldn't get AdityaPandeyCN to show up in the left side drop list of target fork options for some reason but I was able to get to this screen by manipulating the URL.

kdmukai commented:
For example, this PR into a specific fork: https://github.com/KaustubhPatange/seedsigner-emulator/pull/1

Would update that fork owner's PR here: https://github.com/SeedSigner/seedsigner-emulator/pull/1

S1DDHEY commented:
> I couldn't get AdityaPandeyCN to show up in the left side drop list of target fork options for some reason but I was able to get to this screen by manipulating the URL.

was facing the same issue and thought it must be because of permissions issue, should've tried manipulating the URL :smiling_face_with_tear:

AdityaPandeyCN commented:
You need to rebase this PR on top of the latest `dev` branch.

S1DDHEY committed:
Merge branch 'dev' into enhance-l10n-error-handling
